monitoring/prometheus/prometheus.go

```diff
@@ -0,0 +1,115 @@
+
+package prometheus
+
+import (
+	prom "github.com/prometheus/client_golang/prometheus"
+	"time"
+	"google.golang.org/grpc/monitoring"
+	"strings"
+	"google.golang.org/grpc/transport"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	serverStartedCounter = prom.NewCounterVec(
+		prom.CounterOpts{
+			Namespace: "grpc",
+			Subsystem: "server",
+			Name:      "rpc_started_total",
+			Help:      "Total number of RPCs started by the server.",
+		}, []string{"type", "service", "method"})
+
+	serverStreamMsgReceived = prom.NewCounterVec(
+		prom.CounterOpts{
+			Namespace: "grpc",
+			Subsystem: "server",
+			Name:      "rpc_msg_received_total",
+			Help:      "Total number of RPC stream messages received on the server.",
+		}, []string{"type", "service", "method"})
+
+	serverStreamMsgSent = prom.NewCounterVec(
+		prom.CounterOpts{
+			Namespace: "grpc",
+			Subsystem: "server",
+			Name:      "rpc_msg_sent_total",
+			Help:      "Total number of RPC stream messages sent by the server.",
+		}, []string{"type", "service", "method"})
+
+	serverHandledHistogram = prom.NewHistogramVec(
+		prom.HistogramOpts{
+			Namespace: "grpc",
+			Subsystem: "server",
+			Name:      "rpc_handled",
+			Help:      "Histogram of response latency of RPC that had been application-level handled by the server.",
+			Buckets:   prom.DefBuckets,
+		}, []string{"type", "service", "method", "code"})
+
+	serverErred = prom.NewCounterVec(
+		prom.CounterOpts{
+			Namespace: "grpc",
+			Subsystem: "server",
+			Name:      "rpc_erred_total",
+			Help:      "Total number of RPC that had failed on the RPC layer on the server.",
+		}, []string{"type", "service", "method", "error"})
+)
+
+func init() {
+	prom.MustRegister(serverStartedCounter)
+	prom.MustRegister(serverStreamMsgReceived)
+	prom.MustRegister(serverStreamMsgSent)
+	prom.MustRegister(serverHandledHistogram)
+	prom.MustRegister(serverErred)
+}
+
+
+type ServerMonitor struct {
+}
+
+func (m *ServerMonitor) NewServerMonitor(rpcType monitoring.RpcType, fullMethod string) monitoring.RpcMonitor {
+	r := &serverRpcMonitor{rpcType: rpcType, startTime: time.Now()}
+	r.serviceName, r.methodName = splitMethodName(fullMethod)
+	serverStartedCounter.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+	return r
+}
+
+type serverRpcMonitor struct {
```

Comment:
I haven't thought about it, but: the first 3 types are `string`, which is 16 bits each * 3 = 48.
The last item is `time.Time` which is 24 bytes.  So that's 72bits  = 9B total. So there is no padding necessary.
I could turn the `time.Time` into a pointer and reduce it to 56bits, but that is even more wasteful due to pointers.

Does that answer your concerns?